### PR TITLE
[link] feat: add support for external link

### PR DESCRIPTION
**Pull request by maimonav**
Add support for consumers to render their own linkModal( by disabling our modal ), save their data in the linkEntity and dispatch it in viewer when `onClick` event occurs.

### CODE REVIEW

File: e2e/tests/text.e2e.js [hunk 1]
```diff
@@ -132,6 +132,21 @@ describe('text', () => {
     cy.blurEditor();
   });
 
+  it('should insert external link', function() {
+    const testAppConfig = {
+      ...usePluginsConfig({
+        LINK: {
+          isExternalModal: true,
```
Comment (shaulgo) [inline e2e/tests/text.e2e.js:139]: I thought this would be a callback that would allow the consumers to render whatever they need.

Comment (shaulgo) [reply e2e/tests/text.e2e.js:139]: Am I missing something?

Comment (shaulgo) [reply e2e/tests/text.e2e.js:139]: I see you split it to two settings. What do you think about customLinkModal. Also we should probably give it the current data. So user opens it for edit.

**Review by shaulgo — CHANGES_REQUESTED** [review]

File: e2e/test-env/src/client/RicosTestApp/editorPlugins.js [hunk 2]
```diff
@@ -80,6 +80,12 @@ const configs = {
   gallery: {
     scrollingElement: () => window,
   },
+  link: {
+    onLinkAdd: async saveData => {
+      const data = await Promise.resolve({ externalData: {} });
```
[the remark this answers is not in the document]
Comment (maimonav) [reply e2e/test-env/src/client/RicosTestApp/editorPlugins.js:85]: Sure.

File: e2e/test-env/src/client/RicosTestApp/RicosTestApp.jsx [hunk 3]
```diff
@@ -137,7 +138,9 @@ class RicosTestApp extends PureComponent {
         cssOverride={consumerTheme ? consumerTheme : !skipCssOverride && theme}
         seoSettings={seoMode}
         preview={testAppConfig.showDefaultPreview && createPreview()}
-      />
+      >
+        <RichContentViewer config={{ ...testAppConfig.pluginsConfig }} />
```
[the remark this answers is not in the document]
Comment (maimonav) [reply e2e/test-env/src/client/RicosTestApp/RicosTestApp.jsx:142]: It's just for the test-env, the same thing for RichContentEditor.

File: e2e/tests/text.e2e.js (hunk 1 above)
Comment (maimonav) [reply e2e/tests/text.e2e.js:139]: You are right, let me know when we can talk about it, I have some questions :)

Comment (shaulgo) [reply e2e/tests/text.e2e.js:151]: addLink: (saveData) => {
  const data = await renderPanel();
  saveData(data);
}

File: e2e/test-env/src/client/RicosTestApp/RicosTestApp.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (maimonav) [reply e2e/test-env/src/client/RicosTestApp/RicosTestApp.jsx:142]: removed.

File: e2e/test-env/src/client/RicosTestApp/RicosTestApp.jsx [hunk 4]
```diff
@@ -93,12 +93,14 @@ class RicosTestApp extends PureComponent {
           handleFileSelection: mockTestFileUpload,
         };
     const nativeFileUploadConfig = {
-      'wix-draft-plugin-file-upload': uploadHandler,
+      fileUpload: uploadHandler,
     };
 
+    const externalPluginsConfigs = { ...testAppConfig.pluginsConfig, ...nativeFileUploadConfig };
```
Comment (shaulgo) [inline e2e/test-env/src/client/RicosTestApp/RicosTestApp.jsx:99]:
```suggestion
    const customPluginsConfigs = { ...nativeFileUploadConfig, ...testAppConfig.pluginsConfig };
```

File: packages/editor-common/web/src/Utils/draftUtils.ts [hunk 5]
```diff
@@ -50,18 +53,38 @@ export const insertLinkInPosition = (
...
-    url,
-    targetBlank,
-    nofollow,
-    anchorTarget,
-    relValue,
-  });
+  return insertLink(
+    editorState,
+    selection,
+    createLinkEntityData({
```
Comment (shaulgo) [inline packages/editor-common/web/src/Utils/draftUtils.ts:59]: const linkEntityData = ...

File: packages/plugin-link/web/src/toolbar/TextLinkButton.jsx [hunk 6]
```diff
@@ -24,62 +24,73 @@ export default class TextLinkButton extends Component {
...
-      fullScreen: !OriginalLinkPanel,
-      isMobile,
-      customStyles: {
-        content: {
-          position: 'fixed',
+
+    const settings = config[LINK_TYPE];
+    const linkTypes = settings?.linkTypes;
+    const onLinkAdd = settings?.onLinkAdd;
+    const shouldRenderModal = !onLinkAdd;
```
Comment (shaulgo) [inline packages/plugin-link/web/src/toolbar/TextLinkButton.jsx:38]:
```suggestion
    const isExternalLinkHandling = !onLinkAdd;
```

File: packages/plugin-link/web/src/toolbar/TextLinkButton.jsx [hunk 7]
```diff
@@ -24,62 +24,73 @@ export default class TextLinkButton extends Component {
...
+            isMobile,
+            ...commonPanelProps,
+          };
+          helpers.openModal(modalProps);
+        } else {
+          //eslint-disable-next-line no-console
+          console.error(
+            'Open external helper function is not defined for toolbar button with keyName ' +
+              keyName
+          );
```
Comment (shaulgo) [inline packages/plugin-link/web/src/toolbar/TextLinkButton.jsx:80]: put in method

File: packages/plugin-link/web/src/toolbar/TextLinkButton.jsx [hunk 8]
```diff
@@ -24,62 +24,73 @@ export default class TextLinkButton extends Component {
...
     } else {
-      const modalProps = {
-        hidePopup: innerModal.closeInnerModal,
-        top: toolbarOffsetTop,
-        left: toolbarOffsetLeft,
-        modalStyles: OriginalLinkPanel ? null : { maxWidth: 'none', padding: '0 19px' },
-        ...commonPanelProps,
-      };
-      innerModal.openInnerModal(modalProps);
+      onLinkAdd?.(insertExternalLink(getEditorState(), setEditorState));
```
Comment (shaulgo) [inline packages/plugin-link/web/src/toolbar/TextLinkButton.jsx:93]:
```suggestion
      onLinkAdd(insertExternalLink(getEditorState(), setEditorState));
```

Comment (shaulgo) [reply packages/plugin-link/web/src/toolbar/TextLinkButton.jsx:93]: onLinkAdd(text, externlLinkData).then(externlLinkData => setEditorState(insertExternalLink(externlLinkData)))

onLinkAdd(data, externlLinkData, externlLinkData => setEditorState(insertExternalLink(editorState, data))))

Comment (shaulgo) [reply packages/plugin-link/web/src/toolbar/TextLinkButton.jsx:93]: onLinkAdd = (text, externlLinkData) => ... return promise(resolve(data))

onLinkAdd = (text, externlLinkData, cb) => ... cb(data)

**Review by shaulgo — CHANGES_REQUESTED** [review]

File: packages/editor-common/web/src/Utils/draftUtils.ts (hunk 5 above)
Comment (maimonav) [reply packages/editor-common/web/src/Utils/draftUtils.ts:62]: done.

File: packages/plugin-link/web/src/toolbar/TextLinkButton.jsx (hunk 6 above)
Comment (maimonav) [reply packages/plugin-link/web/src/toolbar/TextLinkButton.jsx:38]: done.

File: packages/plugin-link/web/src/toolbar/TextLinkButton.jsx (hunk 7 above)
Comment (maimonav) [reply packages/plugin-link/web/src/toolbar/TextLinkButton.jsx:80]: done.

File: packages/plugin-link/web/src/LinkViewer.jsx [hunk 9]
```diff
@@ -33,15 +33,20 @@ class LinkViewer extends Component {
...
-    this.props?.settings?.onClick?.(event, anchor || this.getHref());
-    if (anchor && !isInEditor) {
-      event.preventDefault();
-      const anchorString = `viewer-${anchor}`;
-      history.pushState({}, null, `#${anchorString}`);
-      const element = document.getElementById(anchorString);
-      anchorScroll(element);
+    const { componentData, isInEditor, settings = {} } = this.props;
+    const { onClick } = settings;
+    if (componentData?.externalData) {
```
Comment (shaulgo) [inline packages/plugin-link/web/src/LinkViewer.jsx:38]: move here event.preventDefault();

File: packages/editor-common/web/src/Utils/draftUtils.ts [hunk 10]
```diff
@@ -52,19 +55,51 @@ export const insertLinkInPosition = (
...
+    const block = contentState.getBlockForKey(blockKey);
+    const entityKey = block.getEntityAt(selection.getStartOffset());
+    const entity = contentState.getEntityMap().get(entityKey);
+    externalLinkData = entity?.getData()?.externalData;
+  }
+
+  return externalLinkData || {};
+};
+
+export const insertExternalLink = (editorState: EditorState, externalData: ExternalLinkData) => {
```
Comment (shaulgo) [inline packages/editor-common/web/src/Utils/draftUtils.ts:85]: ExternalLinkData => any

Comment (maimonav) [reply packages/editor-common/web/src/Utils/draftUtils.ts:85]: Done.

File: packages/plugin-link/web/src/toolbar/TextLinkButton.jsx [hunk 11]
```diff
@@ -11,6 +11,27 @@ import { isEmpty } from 'lodash';
...
+      getExternalLinkData,
+      insertExternalLink,
+      config,
+    } = this.props;
+    const settings = config[LINK_TYPE];
+    const onLinkAdd = settings?.onLinkAdd;
+    const isExternalLinkHandling = onLinkAdd;
+
+    if (isExternalLinkHandling) {
+      const externalLinkData = getExternalLinkData(getEditorState());
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/plugin-link/web/src/toolbar/TextLinkButton.jsx:26]: Done.

**Pull request merged**